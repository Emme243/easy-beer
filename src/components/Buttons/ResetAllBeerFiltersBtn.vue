<template>
  <div>
    <v-tooltip bottom :color="color">
      <template #activator="{ on, attrs }">
        <v-btn
          :color="color"
          :disabled="areFiltersWithDefaultValue"
          @click="resetAllFiltersAndSearchBar"
          depressed
          small
          v-bind="attrs"
          v-on="on"
          data-test-id="reset-all-filters-button"
        >
          <v-icon small color="white">fa-solid fa-filter-circle-xmark</v-icon>
        </v-btn>
      </template>
      <span>Limpiar TODOS los filtros de búsqueda</span>
    </v-tooltip>

    <InfoAlert v-model="isAlertShown" color="red darken-3">
      Se han limpiado TODOS los filtros de búsqueda.
    </InfoAlert>
  </div>
</template>

<script>
import InfoAlert from '@/components/Alerts/InfoAlert';
import { mapActions, mapGetters } from 'vuex';

export default {
  name: 'ResetAllBeerFiltersBtn',
  components: { InfoAlert },
  props: {
    color: {
      type: String,
      default: 'red darken-3',
    },
  },
  data: () => ({ isAlertShown: false }),
  methods: {
    ...mapActions({ resetAllFilters: 'filter/resetAllFilters' }),
    resetAllFiltersAndSearchBar() {
      this.resetAllFilters();
      this.isAlertShown = true;
      setTimeout(() => {
        this.isAlertShown = false;
      }, 3000);
    },
  },
  computed: mapGetters({
    areFiltersWithDefaultValue: 'filter/areFiltersWithDefaultValue',
  }),
};
</script>

<style scoped></style>
