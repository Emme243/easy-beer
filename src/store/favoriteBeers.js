export const favoriteBeersStore = {
  namespaced: true,
  state: {
    favoriteBeers: [],
  },
  mutations: {
    addBeerToFavorites(state, beer) {
      state.favoriteBeers.push(beer);
    },
    removeBeerFromFavorites(state, id) {
      state.favoriteBeers = state.favoriteBeers.filter(beer => beer.id !== id);
    },
    setFavoriteBeers(state, beers) {
      state.favoriteBeers = beers;
    },
    resetFavoriteBeers(state) {
      state.favoriteBeers = [];
    },
  },
  actions: {
    addBeerToFavorites({ commit }, beer) {
      commit('addBeerToFavorites', beer);
    },
    removeBeerFromFavorites({ commit }, beerId) {
      commit('removeBeerFromFavorites', beerId);
    },
    setFavoriteBeers({ commit }, beers) {
      commit('setFavoriteBeers', beers);
    },
    resetFavoriteBeers({ commit }) {
      commit('resetFavoriteBeers');
    },
  },
  getters: {
    getAllFavoriteBeers(state) {
      return state.favoriteBeers;
    },
    isBeerInFavorites(state) {
      return id => {
        return state.favoriteBeers.some(beer => beer.id === id);
      };
    },
  },
};
