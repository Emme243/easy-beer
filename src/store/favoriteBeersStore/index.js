export const favoriteBeersStore = {
  namespaced: true,
  state: {
    favoriteBeers: [],
  },
  getters: {
    getAllFavoriteBeers(state) {
      return state.favoriteBeers;
    },
    isBeerInFavorites(state) {
      return id => {
        return state.favoriteBeers.some(beer => beer.id === id);
      };
    },
  },
  mutations: {
    addBeerToFavorites(state, beer) {
      state.favoriteBeers.push(beer);
    },
    removeBeerFromFavorites(state, id) {
      state.favoriteBeers = state.favoriteBeers.filter(beer => beer.id !== id);
    },
    setFavoriteBeers(state, beers) {
      state.favoriteBeers = beers;
    },
    resetFavoriteBeers(state) {
      state.favoriteBeers = [];
    },
  },
  actions: {
    handleBeerInFavorites({ commit, getters }, beer) {
      const isBeerInFavorites = getters.isBeerInFavorites(beer.id);
      if (isBeerInFavorites) commit('removeBeerFromFavorites', beer.id);
      else commit('addBeerToFavorites', beer);
    },
    setFavoriteBeers({ commit }, beers) {
      commit('setFavoriteBeers', beers);
    },
    resetFavoriteBeers({ commit }) {
      commit('resetFavoriteBeers');
    },
  },
};
