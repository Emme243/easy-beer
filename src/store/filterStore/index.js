import { abvRangeStore } from '@/store/filterStore/abvRangeStore';
import { brewedDateStore } from '@/store/filterStore/brewedDateStore';
import { paginationStore } from '@/store/filterStore/paginationStore';
import { searchQueryStore } from '@/store/filterStore/searchQueryStore';

export const filterStore = {
  namespaced: true,
  modules: {
    abvRange: abvRangeStore,
    brewedDate: brewedDateStore,
    page: paginationStore,
    searchQuery: searchQueryStore,
  },
  state: () => ({}),
  getters: {
    filterStore(_, getters) {
      return {
        minAbvValue: getters.minAbvValue,
        maxAbvValue: getters.maxAbvValue,
        initialBrewedMonth: getters.initialBrewedMonth,
        finalBrewedMonth: getters.finalBrewedMonth,
        page: getters.page,
        searchQuery: getters.searchQuery,
      };
    },
    areFiltersWithDefaultValue(_, getters) {
      return (
        getters.isAbvRangeWithDefaultValue &&
        !getters.initialBrewedMonth &&
        !getters.finalBrewedMonth &&
        !getters.searchQuery
      );
    },
    numberOfFiltersApplied(_, getters) {
      let numberOfFiltersApplied = 0;
      if (!getters.isAbvRangeWithDefaultValue) numberOfFiltersApplied++;
      if (getters.initialBrewedMonth || getters.finalBrewedMonth) numberOfFiltersApplied++;
      return numberOfFiltersApplied;
    },
  },
  mutations: {},
  actions: {
    setAllFilters({ dispatch }, filterQuery) {
      dispatch('setPage', filterQuery.page);
      dispatch('setMinAbvValue', filterQuery.minAbvValue);
      dispatch('setMaxAbvValue', filterQuery.maxAbvValue);
      dispatch('setInitialBrewedMonth', filterQuery.initialBrewedMonth);
      dispatch('setFinalBrewedMonth', filterQuery.finalBrewedMonth);
      dispatch('setSearchQuery', filterQuery.searchQuery);
    },
    resetAllFilters({ dispatch }) {
      dispatch('resetAbvRange');
      dispatch('resetBrewedDate');
      dispatch('setSearchQuery', '');
      dispatch('resetPage');
    },
    resetFiltersInFilterModal({ dispatch }) {
      dispatch('resetAbvRange');
      dispatch('resetBrewedDate');
      dispatch('resetPage');
    },
  },
};
