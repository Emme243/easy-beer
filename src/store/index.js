import { favoriteBeersStore } from '@/store/favoriteBeersStore';
import { filterStore } from '@/store/filterStore';

import Vue from 'vue';
import Vuex from 'vuex';

Vue.use(Vuex);

export default new Vuex.Store({
  modules: {
    filter: filterStore,
    favoriteBeers: favoriteBeersStore,
  },
});
