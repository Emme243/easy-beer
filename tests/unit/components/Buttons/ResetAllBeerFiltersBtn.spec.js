import ResetAllBeerFiltersBtn from '@/components/Buttons/ResetAllBeerFiltersBtn';
import store from '@/store';
import { createLocalVue, mount } from '@vue/test-utils';
import Vuetify from 'vuetify';
import { beersMock } from '../../mocks/beersMock';

describe('Pruebas en ResetAllBeerFiltersBtn.vue', function () {
  let vuetify;
  const localVue = createLocalVue();

  const beer = beersMock[0];
  const mountFunction = options => {
    return mount(ResetAllBeerFiltersBtn, {
      localVue,
      vuetify,
      store,
      propsData: { beer },
      ...options,
    });
  };

  beforeEach(() => {
    vuetify = new Vuetify();
  });

  const resetAllFiltersBtnSelector = '[data-test-id="reset-all-filters-button"]';

  it('Debe hacer match con el snapshot', () => {
    const wrapper = mountFunction();
    expect(wrapper.element).toMatchSnapshot();
  });

  it('El botÃ³n de reseteo debe llamar al dispatch resetAllFilters cuando se hace click y la propiedad isAlertShown debe ser true', async () => {
    store.dispatch = jest.fn();
    const wrapper = mountFunction();
    const resetAllFiltersBtn = wrapper.find(resetAllFiltersBtnSelector);
    expect(resetAllFiltersBtn.exists()).toBe(true);
    resetAllFiltersBtn.vm.$emit('click');
    await wrapper.vm.$nextTick();
    expect(store.dispatch).toHaveBeenCalledWith('filter/resetAllFilters');
    expect(store.dispatch).toHaveBeenCalledTimes(1);
    expect(wrapper.vm.isAlertShown).toBe(true);
  });
});
