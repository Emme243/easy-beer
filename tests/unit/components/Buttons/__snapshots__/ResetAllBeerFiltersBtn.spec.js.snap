// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en ResetAllBeerFiltersBtn.vue Debe hacer match con el snapshot 1`] = `
<div
  beer="[object Object]"
>
  <span
    class="v-tooltip v-tooltip--bottom"
  >
    <!---->
    <button
      aria-expanded="false"
      aria-haspopup="true"
      class="v-btn v-btn--disabled v-btn--has-bg theme--light v-size--small"
      data-test-id="reset-all-filters-button"
      disabled="disabled"
      type="button"
    >
      <span
        class="v-btn__content"
      >
        <i
          aria-hidden="true"
          class="v-icon notranslate fa fa-solid fa-filter-circle-xmark theme--light white--text"
          style="font-size: 16px;"
        />
      </span>
    </button>
  </span>
   
  <div
    class="v-snack v-snack--bottom v-snack--has-background"
  >
    <transition-stub
      name="v-snack-transition"
    >
      <div
        class="v-snack__wrapper v-sheet theme--dark red darken-3"
        style="display: none;"
      >
        <div
          aria-live="polite"
          class="v-snack__content"
          role="status"
        >
           
    Se han limpiado TODOS los filtros de b√∫squeda.
  
        </div>
        <div
          class="v-snack__action "
        >
          <button
            class="v-btn v-btn--outlined v-btn--text theme--dark v-size--default white--text v-snack__btn"
            type="button"
          >
            <span
              class="v-btn__content"
            >
              <i
                aria-hidden="true"
                class="v-icon notranslate fa fa fa fa-times theme--dark"
              />
            </span>
          </button>
        </div>
      </div>
    </transition-stub>
  </div>
</div>
`;
